src/froll.c

```diff
@@ -397,11 +416,207 @@ void frollsumExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool
+      if (ISNAN(x[i])) {
+        truehasnf = true;
+        if (ISNA(x[i])) {
+          iw = i; w = NA_REAL;
+        } else if (ISNA(w)) {
+          // do nothing because w > x[i]: NA > NaN
+        } else {
+          iw = i; w = R_NaN;
+        }
+      } else if (x[i] >= w) {
+        iw = i; w = x[iw];
+      }
+      ans->dbl_v[i] = fill;
+    }
+    if (!truehasnf) { // maybe no NAs
+      for (; i<nx; i++) {
+        if (ISNAN(x[i])) { // Inf properly propagates
+          if (verbose)
+            ansSetMsg(ans, 0, "%s: non-finite values are present in input, continue with extra care for NFs\n", __func__);
+          truehasnf = true;
+          break; // does not increment, w stays as from previous iteration
+        }
+        if (iw+k <= i) { // max left current window
+          iw = i-k; w = R_NegInf;
+          wmax(x, i, k, &w, &iw, true); cmax++;
+        } else if (x[i] >= w) {
+          iw = i; w = x[iw];
+        }
+        ans->dbl_v[i] = w;
+      }
+    }
+    if (truehasnf) {
+      for (; i<nx; i++) { // this loop continues from where "maybe no NAs" loop left
+        if (ISNAN(x[i])) {
+          if (ISNA(x[i])) {
+            iw = i; w = NA_REAL;
+          } else if (ISNA(w)) {
+            // do nothing because w > x[i]: NA > NaN
+          } else { // no NA in window so NaN >= than any non-NA
+            iw = i; w = R_NaN;
+          }
+        } else if (iw+k <= i) { // max left current window
+          iw = i-k; w = R_NegInf;
+          wmax(x, i, k, &w, &iw, false); cmax++;
+        } else if (ISNAN(w)) {
+          // w still within the window and is NA or NaN, x[i] is not NA - already checked above, therefore do nothing
+        } else if (x[i] >= w) {
+          iw = i; w = x[iw];
+        }
+        ans->dbl_v[i] = w;
+      }
+    }
+  }
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: nested window max calculation called %"PRIu64" times\n"), __func__, cmax);
+}
+/* fast rolling max - exact
+ * for each observation in x compute max in window from scratch
+ * faster version ignores NAs (narm=T or has.nf=F), as they are not propagated by `>` operator
+ * otherwise we scan for NaN/NA and run either of two loops
+ * has.nf=FALSE can give incorrect results if NAs provided, documented to be used with care
+ */
+void frollmaxExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool narm, int hasnf, bool verbose) {
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: running in parallel for input length %"PRIu64", window %d, hasnf %d, narm %d\n"), "frollmaxExact", (uint64_t)nx, k, hasnf, (int)narm);
+  for (int i=0; i<k-1; i++) {
+    ans->dbl_v[i] = fill;
+  }
+  if (narm || hasnf==-1) {                                          // ignore NAs as > does not propagate
+    #pragma omp parallel for num_threads(getDTthreads(nx, true))
+    for (uint64_t i=k-1; i<nx; i++) {
+      double w = R_NegInf;
+      for (int j=-k+1; j<=0; j++) {
```

Comment:
Could hint #pragma omp simd reduction(max:w) for this loop  to the compiler

Reply:
I think simd is a new omp hint we would use in DT. Let's keep it as a follow up. It may be useful in few other places probably as well. Filled https://github.com/Rdatatable/data.table/issues/5884